File: note-python/linked-list/queue.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class Queue:
    def __init__(self):
        self.front = None
        self.rear = None

    def enqueue(self, data):
        new_node = Node(data)

        if self.front == None:
            self.front = new_node
            self.rear = new_node
            return

        self.rear.next = new_node
        self.rear = new_node

    def dequeue(self):
        if self.front == None:
            return None
        node = self.front

        if self.front == self.rear:
            self.front = None
            self.rear = None
        else:
            self.front = self.front.next

        return node.data

    def is_empty(self):
        if self.front == None:
            return True
        else:
            return False


if __name__ == "__main__":
    queue = Queue()
    queue.enqueue(1)
    queue.enqueue(2)
    queue.enqueue(3)
    queue.enqueue(4)

    print(queue.dequeue())
    print(queue.dequeue())
    print(queue.dequeue())
    print(queue.dequeue())
    print(queue.dequeue())

File: leetcode/note/queue.js
function Node(data) {
  this.data = data;
  this.next = null;
}

function Queue() {
  let head = null;
  let rear = null;

  this.enqueue = (value) => {
    const newNode = new Node(value);

    if (head === null) {
      head = rear = newNode;
      return;
    }

    rear.next = newNode;
    rear = newNode;
  };

  this.dequeue = () => {
    if (this.empty()) {
      return null;
    }
    const popData = head.data;

    if (head === rear) {
      head = null;
      rear = null;
      return popData;
    }

    head = head.next;

    return popData;
  };

  this.empty = () => {
    return head === null;
  };

  this.reverse = () => {
    let prev = null;
    let next = null;
    let current = head;
    rear = head;

    while (current !== null) {
      next = current.next;
      current.next = prev;
      prev = current;
      current = next;
    }

    head = prev;
  };
}
